{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from flask import Flask, request, send_file\n",
    "from flask_cors import CORS\n",
    "# from flask_ngrok import run_with_ngrok\n",
    "import boto3\n",
    "import pandas as pd\n",
    "from io import BytesIO\n",
    "import matplotlib.pyplot as plt\n",
    "import io\n",
    "import time\n",
    "import mysql.connector as sql\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AWS s3 client configs\n",
    "\n",
    "# s3 = boto3.client('s3',\n",
    "#                   aws_access_key_id='AKIA5UDSK4HA7Y4EHANT',\n",
    "#                   aws_secret_access_key='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',\n",
    "#                   region_name='us-east-1')  # 需要改 #已经改 #已经改"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def load_tpm_datasets(bucket_name): # 函数名要改 #已经改 #已经改\n",
    "#     datasets = []\n",
    "#     object_keys = ['adrenal_dsvm.csv', 'arteryti_dsvm.csv', 'heart_dsvm.csv', 'kidney_dsvm.csv']   # 需要改 #已经改\n",
    "\n",
    "#     for key in object_keys:\n",
    "#         obj = s3.get_object(Bucket=bucket_name, Key=key)\n",
    "#         dataset = pd.read_csv(BytesIO(obj['Body'].read()))\n",
    "#         datasets.append(dataset)\n",
    "\n",
    "#     return datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flask App\n",
    "app = Flask(__name__)\n",
    "CORS(app)\n",
    "# run_with_ngrok(app)\n",
    "\n",
    "# Global variable to store datasets\n",
    "# datasets = []\n",
    "\n",
    "# Load datasets at the start\n",
    "# @app.before_first_request\n",
    "# def load_datasets():\n",
    "#     global datasets\n",
    "#     bucket_name = 'capstone-cre'\n",
    "#     datasets = load_CRE_datasets(bucket_name)\n",
    "\n",
    "# def load_gene_data(gene):\n",
    "#     # Connect to database\n",
    "#     # Database config:\n",
    "#     config = {\n",
    "#         'user': 'root',\n",
    "#         'password': 'zzj1998310',\n",
    "#         'host': 'localhost',\n",
    "#         'database': 'capstone_graph_4',\n",
    "#     }\n",
    "\n",
    "#     # Connect to database\n",
    "#     db = sql.connect(**config)\n",
    "#     cursor = db.cursor()\n",
    "\n",
    "#     # Tables\n",
    "#     query1 = \"SELECT table_name FROM information_schema.tables WHERE table_schema = %s\"\n",
    "#     cursor.execute(query1, (config['database'],))\n",
    "#     tables = [row[0] for row in cursor]\n",
    "\n",
    "#     # Gene data\n",
    "#     return tables, db\n",
    "\n",
    "def plot_tpm_violin(gene):\n",
    "    # Connect to database\n",
    "    # Database config:\n",
    "    config = {\n",
    "        'user': 'root',\n",
    "        'password': 'zzj1998310',\n",
    "        'host': 'localhost',\n",
    "        'database': 'capstone_graph_4',\n",
    "    }\n",
    "\n",
    "    # Connect to database\n",
    "    db = sql.connect(**config)\n",
    "    cursor = db.cursor()\n",
    "\n",
    "    # Tables\n",
    "    query1 = \"SELECT table_name FROM information_schema.tables WHERE table_schema = %s\"\n",
    "    cursor.execute(query1, (config['database'],))\n",
    "    tables = [row[0] for row in cursor]\n",
    "\n",
    "    # Create graphs\n",
    "    plt.figure(figsize=(20, 10))\n",
    "\n",
    "    # Violin plot\n",
    "    all_data = []\n",
    "    positions = []\n",
    "    for i, table in enumerate(tables):\n",
    "        # Query\n",
    "        query = f\"SELECT TPM FROM {table} WHERE description = {gene}\"\n",
    "        df = pd.read_sql(query, db)\n",
    "        if not df.empty:\n",
    "            all_data.append(df['TPM'])\n",
    "            positions.append(i)\n",
    "\n",
    "    # Plot\n",
    "    sns.violinplot(data=all_data, positions=positions, scale='width')\n",
    "    plt.xticks(positions, tables)\n",
    "    plt.xlabel('Table Name')\n",
    "    plt.ylabel('TPM')\n",
    "    plt.title(f\"{gene} in Different Tables\")\n",
    "    plt.tight_layout()\n",
    "\n",
    "\n",
    "\n",
    "@app.route('/generate_tpm_plot', methods=['POST']) # 改名字 #已经改 #已经改\n",
    "def generate_tpm_plot(): # 改名字 #已经改 #已经改\n",
    "    \n",
    "    start_time = time.time()\n",
    "    \n",
    "    data = request.get_json()\n",
    "    \n",
    "    gene = data.get('gene') # 改 #已经改 #已经改\n",
    "\n",
    "    # Call the plot function\n",
    "    plot_tpm_violin(gene) # 改 #已经改 #已经改\n",
    "\n",
    "    # Save the plot to a BytesIO object\n",
    "    buf = io.BytesIO()\n",
    "    plt.savefig(buf, format='jpeg', dpi = 400, bbox_inches = 'tight')\n",
    "    plt.close()\n",
    "    buf.seek(0)\n",
    "    \n",
    "    end_time = time.time()\n",
    "    elapsed_time = end_time - start_time\n",
    "    print(f\"Time taken to generate the image: {elapsed_time} seconds\")\n",
    "\n",
    "    # Send the plot as a response\n",
    "    return send_file(buf, mimetype='image/jpeg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * Serving Flask app '__main__'\n",
      " * Debug mode: off\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.\n",
      " * Running on http://127.0.0.1:7779\n",
      "Press CTRL+C to quit\n",
      "C:\\Users\\13906\\AppData\\Local\\Temp\\ipykernel_16920\\3758575384.py:66: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df = pd.read_sql(query, db)\n",
      "[2023-12-11 02:17:20,665] ERROR in app: Exception on /generate_tpm_plot [POST]\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\pandas\\io\\sql.py\", line 2018, in execute\n",
      "    cur.execute(*args, **kwargs)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\mysql\\connector\\cursor.py\", line 617, in execute\n",
      "    self._handle_result(self._connection.cmd_query(stmt))\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\mysql\\connector\\opentelemetry\\context_propagation.py\", line 77, in wrapper\n",
      "    return method(cnx, *args, **kwargs)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\mysql\\connector\\connection.py\", line 843, in cmd_query\n",
      "    result = self._handle_result(self._send_cmd(ServerCmd.QUERY, query))\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\mysql\\connector\\connection.py\", line 620, in _handle_result\n",
      "    raise get_exception(packet)\n",
      "mysql.connector.errors.ProgrammingError: 1054 (42S22): Unknown column 'CLCN6' in 'where clause'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\flask\\app.py\", line 1455, in wsgi_app\n",
      "    response = self.full_dispatch_request()\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\flask\\app.py\", line 869, in full_dispatch_request\n",
      "    rv = self.handle_user_exception(e)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\flask_cors\\extension.py\", line 176, in wrapped_function\n",
      "    return cors_after_request(app.make_response(f(*args, **kwargs)))\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\flask\\app.py\", line 867, in full_dispatch_request\n",
      "    rv = self.dispatch_request()\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\flask\\app.py\", line 852, in dispatch_request\n",
      "    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Temp\\ipykernel_16920\\3758575384.py\", line 91, in generate_tpm_plot\n",
      "    plot_tpm_violin(gene) # 改 #已经改 #已经改\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Temp\\ipykernel_16920\\3758575384.py\", line 66, in plot_tpm_violin\n",
      "    df = pd.read_sql(query, db)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\pandas\\io\\sql.py\", line 564, in read_sql\n",
      "    return pandas_sql.read_query(\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\pandas\\io\\sql.py\", line 2078, in read_query\n",
      "    cursor = self.execute(*args)\n",
      "  File \"C:\\Users\\13906\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\pandas\\io\\sql.py\", line 2030, in execute\n",
      "    raise ex from exc\n",
      "pandas.errors.DatabaseError: Execution failed on sql 'SELECT TPM FROM adrenal_gland WHERE description = CLCN6': 1054 (42S22): Unknown column 'CLCN6' in 'where clause'\n",
      "127.0.0.1 - - [11/Dec/2023 02:17:20] \"POST /generate_tpm_plot HTTP/1.1\" 500 -\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 2000x1000 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Run the Flask App\n",
    "if __name__ == '__main__':\n",
    "    app.run(port=7779)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
